### ISSUE TITLE
Inference Engine supported model?

### ISSUE THREAD

**Issue by hyunwoongko** [microsoft/DeepSpeed#1180]
May I know which models currently support model parallelization in inference? As a result of the verification, some models appear to replicate parameters as-is to other devices.

**Comment by hyunwoongko** [comment]
https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/module_inject/replace_policy.py

Are only the above 4 models currently supported?

**Comment by RezaYazdaniAminabadi** [comment]
Hi @hyunwoongko 

This is true that we need to have replace policies for different models in order to both correctly parallelize and inject the kernels.
One caveat though is that it always needs to run through the DeepSpeed Inference API which knows where the results across different GPUs can be merged. Currently we support encoder-only or decoder-only architectures. I am working on the remaining pieces to add in the case of encoder-decoder models.
Thanks,
Reza

**Comment by hyunwoongko** [comment]
https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/module_inject/replace_module.py#L152

By the way, why doesn't qkvb change to fp16 in this part?

**Comment by RezaYazdaniAminabadi** [comment]
It might be None. I just moved it here: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/module_inject/replace_module.py#L234

**Comment by hyunwoongko** [comment]
It looks to me like it's for GPTNeo. Thank you.
Could you tell me when the Inference Engine of the Seq2Seq architecture will be released?

**Comment by hyunwoongko** [comment]
@RezaYazdaniAminabadi 
Could you tell me when the Inference Engine of the Seq2Seq architecture will be released?

**Comment by RezaYazdaniAminabadi** [comment]
Hi @hyunwoongko 

Sorry for the late reply.
I am not exactly sure about the time this will be ready, but I would say this is not going to take that long. I am just looking into different models and understand their differences to find a generic way of supporting them that does not require too many changes for the inference engine. Can I ask for a few models that you have in mind regarding the seq2seq architecture?
Thanks,
Reza

**Comment by hyunwoongko** [comment]
I've been working on implementing a new inference layer about seq2seq architectures for a deep speed inference engine for a few days. The problem is that in the case of the seq2seq model, there were quite a few models with various types of attention masks or inputs, and it was not easy to generalize them. As an example, T5 and Bart have very different structures, and models like Funnel Transformer have very unique structures, so it was not easy to parallelize them in current way. (covering all models with one DeepSpeedInference model seems to be a fairly difficult task.) 
<br>

> Can I ask for a few models that you have in mind regarding the seq2seq architecture?

First of all, it would be good if Bart, the most basic of the Pre-trained Seq2Seq model, and models that follow Bart's architecture would be implemented. (Actually many Seq2Seq models such as BlenderBot was implemented by copying Bart's code.)
<br>

I'm currently working on deploying quite large models (10 ~ 30+ Billion) for a company and I've been trying to use the DeepSpeed ​​Inference Engine for this. The DeepSpeed ​​Inference Engine is very creative and really great, but there are a few things that need to be improved before it can actually be used for web server deployment. So, I'm implementing a whole new open source framework myself for model parallel deployment, borrowing some ideas from DeepSpeed. I hope to releasing this soon, so we can share more ideas and contribute better things to the community together.


Thanks :)
Hyunwoong Ko.
